File: src/lib/charts/adapters/atlas.ts
import { Chart, ChartCategory, ChartProvider } from '../types';

export class AtlasVACAdapter implements ChartProvider {
  name = 'Atlas VAC';
  
  constructor() {}

  async getCharts(icao: string): Promise<Chart[]> {
    const cycleName = process.env.NEXT_PUBLIC_AIRAC_CYCLE_NAME || 'eAIP_22_JAN_2026';
    // URL pattern: .../Atlas-VAC/PDF_AIPparSSection/VAC/AD/AD-2.<ICAO>.pdf
    const url = `https://www.sia.aviation-civile.gouv.fr/media/dvd/${cycleName}/Atlas-VAC/PDF_AIPparSSection/VAC/AD/AD-2.${icao}.pdf`;

    try {
      // Check for existence using HEAD request to avoid downloading the full PDF
      const response = await fetch(url, { method: 'HEAD' });
      
      if (!response.ok) {
        if (response.status === 404) return [];
        // For other errors, we might want to log or ignore. 
        // If the service is down, maybe return empty?
        console.warn(`Atlas VAC check failed for ${icao}: ${response.status}`);
        return [];
      }

      // If exists, return the chart
      return [{
        id: url,
        source: 'ATLAS',
        category: ChartCategory.VAC,
        subtitle: 'Carte VAC',
        filename: `AD-2.${icao}.pdf`,
        url: url,
        tags: ['VAC']
      }];

    } catch (error) {
      console.error('Atlas VAC Adapter Error:', error);
      return [];
    }
  }
}

File: crates/atc-book/src/adapters/atlas.rs
use crate::airac::AiracCycle;
use crate::models::{Chart, ChartCategory, ChartSource};

/// Check if an Atlas VAC PDF exists for a given ICAO code.
pub async fn fetch_charts(icao: &str, airac: &AiracCycle) -> Result<Vec<Chart>, String> {
    let base = format!(
        "https://www.sia.aviation-civile.gouv.fr/media/dvd/{}/FRANCE/{}/html/eAIP",
        airac.sia_cycle_name(),
        airac.sia_airac_date(),
    );
    let url = format!(
        "{}/Cartes/{}/AD-2.{}.pdf",
        base,
        icao.to_uppercase(),
        icao.to_uppercase()
    );
    let provider_relative_url = format!(
        "Cartes/{}/AD-2.{}.pdf",
        icao.to_uppercase(),
        icao.to_uppercase()
    );

    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .build()
        .map_err(|e| format!("HTTP client error: {e}"))?;

    let resp = client
        .head(&url)
        .send()
        .await
        .map_err(|e| format!("Atlas request failed: {e}"))?;

    if resp.status().is_success() || resp.status().as_u16() == 302 {
        let id = format!("{}-ATLAS-VAC", icao.to_uppercase());
        Ok(vec![Chart {
            id,
            source: ChartSource::Atlas,
            category: ChartCategory::Vac,
            subtitle: format!("Atlas VAC {}", icao.to_uppercase()),
            filename: format!("AD-2.{}.pdf", icao.to_uppercase()),
            provider_relative_url,
            linked_provider_relative_urls: Vec::new(),
            airac_code: airac.code.clone(),
            page: None,
            tags: vec!["VAC".to_string()],
            runways: Vec::new(),
            custom_title: None,
        }])
    } else {
        Ok(Vec::new())
    }
}
